firewood/src/storage/linear/proposed.rs

```diff
@@ -0,0 +1,526 @@
+    // TODO: we might be able to optimize the case where we keep adding data to
+    // the end of the file by not coalescing left. We'd have to change the assumption
+    // in the reader that when you reach the end of a modified region, you're always
+    // in an unmodified region
+
+    fn write(&mut self, offset: u64, object: &[u8]) -> Result<usize, Error> {
+        // the structure of what will eventually be inserted
+        struct InsertData {
+            offset: u64,
+            data: Box<[u8]>,
+        }
+
+        debug_assert!(!object.is_empty());
+        let mut insert = InsertData {
+            offset,
+            data: Box::from(object),
+        };
+
+        // we track areas to delete here, which happens when we combine
+        let mut merged_offsets_to_delete = Vec::new();
+
+        // TODO: combining a change where we constantly append to a file might be too
+        // expensive, so consider skipping the relocations if they don't overlap at all
+        for existing in self.new.range(..=insert.offset).rev() {
+            let old_end = *existing.0 + existing.1.len() as u64;
+            // found    smme--------- s=existing.0 e=old_end
+            // original ----smmme---- s=new_start e=new_start+object.len() [0 overlap]
+            // original --smmmme----- with overlap
+            // new      smmmmmmme----
+            if insert.offset <= old_end {
+                // we have some overlap, so we'll have only one area when we're done
+                // mark the original area for deletion and recompute it
+                // TODO: can this be done with iter/zip?
+                // TODO: can we avoid copying the mutating data?
+                let mut new_data = vec![];
+                for resulting_bytes_offset in *existing.0.. {
+                    new_data.push(
+                        if resulting_bytes_offset >= insert.offset
+                            && resulting_bytes_offset < insert.offset + insert.data.len() as u64
+                        {
+                            *insert
+                                .data
+                                .get((resulting_bytes_offset - insert.offset) as usize)
+                                .expect("bounds checked above")
+                        } else if let Some(res) = existing
+                            .1
+                            .get((resulting_bytes_offset - *existing.0) as usize)
+                        {
+                            *res
+                        } else {
+                            break;
+                        },
+                    );
+                }
+                insert.offset = *existing.0;
+                merged_offsets_to_delete.push(insert.offset);
+                insert.data = new_data.into_boxed_slice();
+            }
+        }
+        // TODO: search for modifications after this one that can be merged
+        for existing in self.new.range(insert.offset + 1..) {
```

Comment:
Same question as above -- does this loop need to iterate more than once?

Reply:
this one definitely does! Imagine you have two changes:
 `omomo`
and you're writing over both of them:
 `MMMM-`
we want to do this loop twice in that case to consume both original modifications.